waku/common/postgres/asyncpool.nim

```diff
@@ -0,0 +1,176 @@
+# Simple async pool driver for postgress.
+# Inspired by: https://github.com/treeform/pg/
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/sequtils,
+  stew/results,
+  chronicles,
+  chronos
+import
+  ./common,
+  ./connection
+
+logScope:
+  topics = "postgres asyncpool"
+
+
+## Database connection pool options
+
+type PgAsyncPoolOptions* = object
+    minConnections: int
+    maxConnections: int
+
+func init*(T: type PgAsyncPoolOptions, minConnections: Positive = 1, maxConnections: Positive = 5): T =
+  if minConnections > maxConnections:
+    raise newException(Defect, "maxConnections must be greater or equal to minConnections")
+
+  PgAsyncPoolOptions(
+    minConnections: minConnections,
+    maxConnections: maxConnections
+  )
+
+func minConnections*(options: PgAsyncPoolOptions): int =
+  options.minConnections
+
+func maxConnections*(options: PgAsyncPoolOptions): int =
+  options.maxConnections
+
+
+## Database connection pool
+
+type PgAsyncPoolState {.pure.} = enum
+    Live,
+    Closing,
+    Closed
+
+type
+  ## Database connection pool
+  PgAsyncPool* = ref object
+    connOptions: PgConnOptions
+    poolOptions: PgAsyncPoolOptions
+
+    state: PgAsyncPoolState
+    conns: seq[DbConn]
```

Comment:
Personally I would have created a new type to hold the DbConn & bool instead of two seq

Reply:
It's another possibility, yes 🙂 Some parts come from _treeforms_'s implementation. This is one of them.